File: benchmarks/merge_sorted_lists.c

#include <stdio.h>
#include <stdlib.h>

struct ListNode {
  int val;
  struct ListNode *next;
};

struct ListNode *create_list(int len, int start, int step) {
  if (len <= 0)
    return NULL;
  struct ListNode *node = (struct ListNode *)malloc(sizeof(struct ListNode));
  node->val = start;
  node->next = create_list(len - 1, start + step, step);
  return node;
}

void free_list(struct ListNode *head) {
  if (!head)
    return;
  free_list(head->next);
  free(head);
}

struct ListNode *merge_two_lists(struct ListNode *l1, struct ListNode *l2) {
  if (!l1)
    return l2;
  if (!l2)
    return l1;

  if (l1->val < l2->val) {
    l1->next = merge_two_lists(l1->next, l2);
    return l1;
  } else {
    l2->next = merge_two_lists(l1, l2->next);
    return l2;
  }
}

int main() {
  int checksum = 0;

  for (int i = 0; i < 5000; i++) {
    struct ListNode *l1 = create_list(100, 0, 2);
    struct ListNode *l2 = create_list(100, 1, 2);
    struct ListNode *merged = merge_two_lists(l1, l2);

    struct ListNode *curr = merged;
    while (curr) {
      checksum += curr->val;
      curr = curr->next;
    }

    free_list(merged);
  }

  printf("Checksum: %d\n", checksum);
  return 0;
}

File: benchmarks/merge_sorted_lists.rs

struct ListNode {
    val: i32,
    next: Option<Box<ListNode>>,
}

impl ListNode {
    fn new(val: i32) -> Self {
        ListNode { val, next: None }
    }
}

fn create_list(len: i32, start: i32, step: i32) -> Option<Box<ListNode>> {
    if len <= 0 {
        return None;
    }
    let mut node = ListNode::new(start);
    node.next = create_list(len - 1, start + step, step);
    Some(Box::new(node))
}

fn merge_two_lists(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    match (l1, l2) {
        (None, None) => None,
        (Some(n), None) => Some(n),
        (None, Some(n)) => Some(n),
        (Some(mut n1), Some(mut n2)) => {
            if n1.val < n2.val {
                n1.next = merge_two_lists(n1.next, Some(n2));
                Some(n1)
            } else {
                n2.next = merge_two_lists(Some(n1), n2.next);
                Some(n2)
            }
        }
    }
}

fn main() {
    let mut checksum = 0;
    for _ in 0..5000 {
        let l1 = create_list(100, 0, 2);
        let l2 = create_list(100, 1, 2);
        let merged = merge_two_lists(l1, l2);
        
        let mut curr = &merged;
        while let Some(node) = curr {
            checksum += node.val;
            curr = &node.next;
        }
    }
    println!("Checksum: {}", checksum);
    // std::hint::black_box(checksum);
}
